### ISSUE TITLE
Fn::If parameters not being evaluated properly

### ISSUE THREAD

**Issue by eemperor** [martysweet/cfn-lint#106]
cfn-lint is reporting critical errors when processing the following template:

[Cfn Template](https://github.com/plus3it/terraform-aws-watchmaker/blob/master/modules/lx-instance/watchmaker-lx-instance.cfn.json)

[Affected Section, Line 1049](https://github.com/plus3it/terraform-aws-watchmaker/blob/master/modules/lx-instance/watchmaker-lx-instance.cfn.json#L1049)

The template above functions properly when used to launch an instance in AWS.

The critical errors for both conditional states in `Fn::If` are shown below:

`False` Condition (default)
```
$ cfn-lint validate watchmaker-lx-instance.cfn.json --parameters Name="test123lxinst",AmiId="ami-1c8ee466",AmiDistro="CentOS"
0 infos
0 warn
1 crit
Resource: Resources > WatchmakerInstance > Properties > BlockDeviceMappings > 1
Message: Expecting an object, got undefined
Documentation: http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-instance.html

Template invalid!
```

`True` Condition
```
$ cfn-lint validate watchmaker-lx-instance.cfn.json --parameters Name="test123lxinst",AmiId="ami-1c8ee466",AmiDistro="CentOS",AppVolumeDevice="true"
0 infos
0 warn
2 crit
Resource: Resources > WatchmakerInstance > Properties > BlockDeviceMappings > 1
Message: Fn::If does not allow DeviceName as a nested function
Documentation: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-conditions.html#intrinsic-function-reference-conditions-if

Resource: Resources > WatchmakerInstance > Properties > BlockDeviceMappings > 1
Message: Expecting an object, got 'INVALID_IF_STATEMENT'
Documentation: http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-instance.html

Template invalid!
```

It may be related to a previously closed [issue](https://github.com/martysweet/cfn-lint/issues/61).

**Comment by eemperor** [comment]
Here's a simplified template that exhibits the issue:
```
{
    "AWSTemplateFormatVersion" : "2010-09-09",
    "Description" : "This is a test template.  DO NOT USE!",
    "Parameters" :
    {
        "AmiId" :
        {
            "Description" : "ID of the AMI to launch",
            "Type" : "String",
            "AllowedPattern" : "^ami-[0-9a-z]{8}$|^ami-[0-9a-z]{17}$"
        },
        "AmiDistro" :
        {
            "Description" : "Linux distro of the AMI",
            "Type" : "String",
            "AllowedValues" :
            [
                "AmazonLinux",
                "CentOS",
                "RedHat"
            ]
        },
        "AppVolumeDevice" :
        {
            "Description" : "Decision on whether to mount an extra EBS volume. Leave as default (\"false\") to launch without an extra application volume",
            "Type" : "String",
            "Default" : "false",
            "AllowedValues" :
            [
                "true",
                "false"
            ]
        },
        "AppVolumeType" :
        {
            "Description" : "Type of EBS volume to create. Ignored if \"AppVolumeDevice\" is false",
            "Type" : "String",
            "Default" : "gp2",
            "AllowedValues" :
            [
                "gp2",
                "io1",
                "sc1",
                "st1",
                "standard"
            ]
        },
        "AppVolumeSize" :
        {
            "Description" : "Size in GB of the EBS volume to create. Ignored if \"AppVolumeDevice\" is false",
            "Type" : "Number",
            "Default" : "1",
            "MinValue": "1",
            "MaxValue": "16384",
            "ConstraintDescription" : "Must be between 1GB and 16384GB."
        },
        "InstanceType" :
        {
            "Description" : "Amazon EC2 instance type",
            "Type" : "String",
            "Default" : "t2.micro",
            "AllowedValues" :
            [
                "t2.micro",
                "t2.small",
                "t2.medium",
                "t2.large",
                "t2.xlarge",
                "c4.large",
                "c4.xlarge",
                "m4.large",
                "m4.xlarge",
                "c5.large",
                "c5.xlarge",
                "m5.large",
                "m5.xlarge"
            ]
        }
    },
    "Conditions" :
    {
        "CreateAppVolume" :
        {
            "Fn::Equals" : [ { "Ref" : "AppVolumeDevice" }, "true" ]
        }
    },
    "Mappings" :
    {
        "Distro2RootDevice" :
        {
            "AmazonLinux" : { "DeviceName" : "xvda" },
            "RedHat" : { "DeviceName" : "sda1" },
            "CentOS" : { "DeviceName" : "sda1" }
        },
        "InstanceTypeCapabilities" :
        {
            "t2.micro" :
            {
                "ExternDeviceName" : "/dev/xvdf",
                "InternDeviceName" : "/dev/xvdf"
            },
            "t2.small" :
            {
                "ExternDeviceName" : "/dev/xvdf",
                "InternDeviceName" : "/dev/xvdf"
            },
            "t2.medium" :
            {
                "ExternDeviceName" : "/dev/xvdf",
                "InternDeviceName" : "/dev/xvdf"
            },
            "t2.large" :
            {
                "ExternDeviceName" : "/dev/xvdf",
                "InternDeviceName" : "/dev/xvdf"
            },
            "t2.xlarge" :
            {
                "ExternDeviceName" : "/dev/xvdf",
                "InternDeviceName" : "/dev/xvdf"
            },
            "c4.large" :
            {
                "ExternDeviceName" : "/dev/xvdf",
                "InternDeviceName" : "/dev/xvdf"
            },
            "c4.xlarge" :
            {
                "ExternDeviceName" : "/dev/xvdf",
                "InternDeviceName" : "/dev/xvdf"
            },
            "m4.large" :
            {
                "ExternDeviceName" : "/dev/xvdf",
                "InternDeviceName" : "/dev/xvdf"
            },
            "m4.xlarge" :
            {
                "ExternDeviceName" : "/dev/xvdf",
                "InternDeviceName" : "/dev/xvdf"
            },
            "c5.large" :
            {
                "ExternDeviceName" : "/dev/xvdf",
                "InternDeviceName" : "/dev/nvme1n1"
            },
            "c5.xlarge" :
            {
                "ExternDeviceName" : "/dev/xvdf",
                "InternDeviceName" : "/dev/nvme1n1"
            },
            "m5.large" :
            {
                "ExternDeviceName" : "/dev/xvdf",
                "InternDeviceName" : "/dev/nvme1n1"
            },
            "m5.xlarge" :
            {
                "ExternDeviceName" : "/dev/xvdf",
                "InternDeviceName" : "/dev/nvme1n1"
            }
        }
    },
    "Resources" :
    {
        "WatchmakerInstance" :
        {
            "Type" : "AWS::EC2::Instance",
            "CreationPolicy" :
            {
                "ResourceSignal" :
                {
                    "Count" : "1",
                    "Timeout" : "PT30M"
                }
            },
            "Properties" :
            {
                "ImageId" : { "Ref" : "AmiId" },
                "InstanceType" : { "Ref" : "InstanceType" },
                "BlockDeviceMappings" :
                [
                    {
                        "DeviceName" :
                        { "Fn::Join" : [ "", [
                            "/dev/",
                            { "Fn::FindInMap" :
                                [
                                    "Distro2RootDevice",
                                    { "Ref" : "AmiDistro" },
                                    "DeviceName"
                                ]
                            }
                        ]]},
                        "Ebs" :
                        {
                            "VolumeType" : "gp2",
                            "DeleteOnTermination" : "true"
                        }
                    },
                    {
                        "Fn::If" :
                        [
                            "CreateAppVolume",
                            {
                                "DeviceName" :
                                {
                                    "Fn::FindInMap" :
                                    [
                                        "InstanceTypeCapabilities",
                                        { "Ref" : "InstanceType"},
                                        "ExternDeviceName"
                                    ]
                                },
                                "Ebs" :
                                {
                                    "VolumeSize" : { "Ref" : "AppVolumeSize" },
                                    "VolumeType" : { "Ref" : "AppVolumeType" },
                                    "DeleteOnTermination" : "true"
                                }
                            },
                            { "Ref" : "AWS::NoValue" }
                        ]
                    }
                ],
                "UserData" :
                {
                   "Ref" : "AWS::NoValue"
                }
            }
        }
    },
    "Outputs" :
    {
        "WatchmakerInstanceId" :
        {
            "Value" : { "Ref" : "WatchmakerInstance" },
            "Description" : "Instance ID"
        }
    }
}
```

**Comment by martysweet** [comment]
Thanks @eemperor, it looks like there are two slightly different problems here.

The first is the correct operation of the Fn::If statement, however, the Ref of AWS::NoValue results in undefined, as specified here. https://github.com/martysweet/cfn-lint/blob/f182abdbe58aca035450ed875fc6206d8b9f20d9/src/awsData.ts#L118 

This should be handled too (I assume) null out the property defined. This may be as simple as doing a resolveIntrinsicFunction https://github.com/martysweet/cfn-lint/blob/master/src/validator.ts#L799 and if it's undefined return null? @akdor1154 might know more on this side of things.

The second is the incorrectly assumed object validation at https://github.com/martysweet/cfn-lint/blob/master/src/validator.ts#L796-L802, where an object is being received and it is trying to resolve it into an intrinsic function instead of passing the object up by returning it. In this case, it might just be a case of "if not an intrinsic function" return the value given. However, I can see potential issues with the lack of recursion into those parts of the templates if we do that.

I should be able to look further into this next week, feel free to have a go at fixing it and make a PR :1st_place_medal: 

Marty

**Comment by mijndert** [comment]
Any word on this?

**Comment by martysweet** [comment]
@mijndert Can you provide the most simple test case which makes it break? (<15 lines)

@RazzM13 A nice brain teaser for you :smiley:

**Comment by mijndert** [comment]
@martysweet sure

My Parameter:

```HTTPSListener:                {Description: HTTPS Listener,           Type: String,   AllowedValues: ["True", "False"]}```

My Condition:

```CreateHTTPSListener:  !Equals [ !Ref HTTPSListener, False ]```

And my actual resource:

```
ListenerHTTPS:
    Type: AWS::ElasticLoadBalancingV2::Listener
    Condition: CreateHTTPSListener
    Properties:
      DefaultActions:
      - Type: forward
        TargetGroupArn: !Ref TargetGroup
      LoadBalancerArn: !Ref LoadBalancer
      Port: '443'
      Protocol: HTTPS
      Certificates:
      - CertificateArn: !If [CreateHTTPSCert, !Ref CertificateHTTPS, !Ref SSLCertARN]
```

**Comment by martysweet** [comment]
Might be fixed :+1: Can you test it out from the `fix-condition-issue-106` branch?

**Comment by martysweet** [comment]
This should now be fixed in v1.7.4, can you test it out?

**Comment by eemperor** [comment]
Looks like the update fixed this issue.  Thank you.  👍
